rustls/src/server/hs.rs

```diff
@@ -239,52 +233,63 @@ impl ExtensionProcessing {
-            extension_type: ExtensionType::ClientCertificateType,
-        };
         self.process_cert_type_extension(
-            raw_key_negotation_params.validate_raw_key_negotiation(),
+            client_supports,
+            config
+                .verifier
+                .requires_raw_public_keys(),
+            ExtensionType::ClientCertificateType,
             cx,
         )
     }
 
     fn process_cert_type_extension(
         &mut self,
-        raw_key_negotiation_result: RawKeyNegotationResult,
+        client_supports: Vec<CertificateType>,
+        requires_raw_keys: bool,
+        extension_type: ExtensionType,
         cx: &mut ServerContext<'_>,
     ) -> Result<(), Error> {
-        match raw_key_negotiation_result {
-            RawKeyNegotationResult::Negotiated(ExtensionType::ClientCertificateType) => {
+        let raw_key_negotation_result = match (
+            requires_raw_keys,
+            client_supports.contains(&CertificateType::RawPublicKey),
+            client_supports.contains(&CertificateType::X509),
+        ) {
+            (true, true, _) => Ok((extension_type, CertificateType::RawPublicKey)),
+            (false, _, true) => Ok((extension_type, CertificateType::X509)),
+            (false, true, false) => Err(Error::PeerIncompatible(
+                PeerIncompatible::IncorrectCertificateTypeExtension,
+            )),
+            (true, false, _) => Err(Error::PeerIncompatible(
+                PeerIncompatible::IncorrectCertificateTypeExtension,
+            )),
+            (false, false, false) => return Ok(()),
+        };
+
+        match raw_key_negotation_result {
+            Ok((ExtensionType::ClientCertificateType, cert_type)) => {
                 self.exts
                     .push(ServerExtension::ClientCertType(
-                        CertificateType::RawPublicKey,
+                        cert_type
                     ));
             }
-            RawKeyNegotationResult::Negotiated(ExtensionType::ServerCertificateType) => {
+            Ok((ExtensionType::ServerCertificateType, cert_type)) => {
                 self.exts
                     .push(ServerExtension::ServerCertType(
-                        CertificateType::RawPublicKey,
+                        cert_type
                     ));
             }
-            RawKeyNegotationResult::Err(err) => {
+            Err(err) => {
                 return Err(cx
                     .common
                     .send_fatal_alert(AlertDescription::HandshakeFailure, err));
             }
-            RawKeyNegotationResult::NotNegotiated => {}
-            RawKeyNegotationResult::Negotiated(_) => unreachable!(
-                "The extension type should only ever be ClientCertificateType or ServerCertificateType"
-            ),
+            Ok((_,_)) => return Err(Error::General("The extension type should only ever be ClientCertificateType or ServerCertificateType".to_string())),
```

Comment:
I would like to avoid new uses of "Error::General" -- in this case I think this statement is statically unreachable. I would suggest this instead:

- `Ok((_,_)) => unreachable!(),` here.
- `debug_assert!(extension_type == ExtensionType::ClientCertificateType || extension_type == ExtensionType::ServerCertificateType);` at the top of this function to record its valid inputs
- then drop the artificial test introduced at the bottom